### PR TITLE
ARROW-2500: [Java] IPC Writers/readers are not always setting validity bits correctly

### PR DESCRIPTION
**Pull request by bomeng**
Fix the issue of getNullCount() does not return the correct result in certain cases. 

For example, validityBuffer is 0b10110, while valueCount=3, the null count should be 1, but currently it returns 0.

Fix approach:
1. Based on the valueCount, modify the last byte by mask the remaining bits to be all 1's.
0b10110 will become 0b11111110
2. Count how many 1 in the byte by using bitCount()
3. Use 8 * sizeInBytes - count to get the total 0's

Test: 
Added 2 tests to the existing test classes;
Created 1 new file to purposely test BitVectorHelper.java, since it has some public static method and we may add more testes in the future.

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 1]
```diff
@@ -140,22 +140,18 @@ public static int getNullCount(final ArrowBuf validityBuffer, final int valueCou
...
-       */
-      count += Integer.bitCount(byteValue & 0x7F) - (byteValue >> 7);
+    for (int i = 0; i < sizeInBytes; i++) {
+      byte byteValue = validityBuffer.getByte(i);
+      if (i == sizeInBytes - 1 && remainder != 0) {
+        // making the remaining bits all 1s
+        byte mask = (byte)(0xFF << remainder);
+        byteValue = (byte) (byteValue | mask);
+      }
+      count += Integer.bitCount(byteValue & 0xFF);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java]: This looks ok by me because it will cast up to an integer and only keep the 8 original bits, I do think we need better test coverage for this though

File: java/vector/src/test/java/org/apache/arrow/vector/TestBitVectorHelper.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import io.netty.buffer.ArrowBuf;
+import io.netty.buffer.PooledByteBufAllocatorL;
+
+public class TestBitVectorHelper {
+    @Test
+    public void testGetNullCount() throws Exception {
+        ArrowBuf validityBuffer = new ArrowBuf(
+                null, null, new PooledByteBufAllocatorL().empty,
+                null, null, 0, 3, true);
+        validityBuffer.setByte(0, 22);
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestBitVectorHelper.java]: could you add some comments on this test, since it is not straightforward that you have a value count of 3, but have bits set beyond that?

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java [hunk 3]
```diff
@@ -708,6 +711,56 @@ public void testWriteReadVarBin() throws IOException {
...
+        vector.setValueCount(5);
+        root.setRowCount(5);
+        writer.writeBatch();
+
+        vector.setNull(0);
+        vector.setSafe(1, 1);
+        vector.setSafe(2, 2);
+        vector.setValueCount(3);
+        root.setRowCount(3);
+        writer.writeBatch();
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: You should be able to put the body of this in BaseFileTest and then use the same thing for ArrowFileReader and ArrowStreamReader

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java [hunk 4]
```diff
@@ -99,4 +99,56 @@ public void testReadWrite() throws IOException {
...
+    ByteArrayInputStream in = new ByteArrayInputStream(os.toByteArray());
+
+    try (ArrowStreamReader reader = new ArrowStreamReader(in, allocator);) {
+      IntVector read = (IntVector) reader.getVectorSchemaRoot().getFieldVectors().get(0);
+
+      reader.loadNextBatch();
+
+      assertEquals(read.getValueCount(), 5);
+      assertNull(read.getObject(0));
+      assertEquals(read.getObject(1), Integer.valueOf(1));
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java]: you can just use the `get` method that returns an integer

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java [hunk 5]
```diff
@@ -99,4 +99,56 @@ public void testReadWrite() throws IOException {
...
+
+    ByteArrayInputStream in = new ByteArrayInputStream(os.toByteArray());
+
+    try (ArrowStreamReader reader = new ArrowStreamReader(in, allocator);) {
+      IntVector read = (IntVector) reader.getVectorSchemaRoot().getFieldVectors().get(0);
+
+      reader.loadNextBatch();
+
+      assertEquals(read.getValueCount(), 5);
+      assertNull(read.getObject(0));
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java]: use `isNull` method instead of `getObject`

**Review by BryanCutler — CHANGES_REQUESTED** [review]
Thanks @bomeng for the fix and adding some tests! I think this could be optimized a little more and just a few minor things to cleanup.

**Comment by BryanCutler** [comment]
LGTM pending CI, would like to check with @elahrvivaz @siddharthteotia @icexelloss for any other comments

**Review by BryanCutler — APPROVED** [review]

**Review by elahrvivaz — APPROVED** [review]
I tested our code with this branch, and didn't see any issues. Thanks!

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java (hunk 4 above)
Comment (elahrvivaz) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java]: these were from my original test - we use getObject as we don't care about the runtime type of the value, and we need to wrap the primitive anyway.

**Comment by BryanCutler** [comment]
Thanks for testing @elahrvivaz!  I'll commit this later today if no more comments.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java [hunk 6]
```diff
@@ -18,6 +18,8 @@
 
 package org.apache.arrow.vector.ipc;
 
+import static org.junit.Assert.assertEquals;
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: I don't think this is used anymore

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java [hunk 7]
```diff
@@ -26,23 +28,23 @@
...
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
-
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: nit: unrelated change

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java [hunk 8]
```diff
@@ -708,6 +710,25 @@ public void testWriteReadVarBin() throws IOException {
...
 
+  @Test
+  public void testReadWriteMultipleBatches() throws IOException {
+    File file = new File("target/mytest_nulls_multibatch.arrow");
+
+    try (IntVector vector = new IntVector("foo", allocator);) {
+      Schema schema = new Schema(Collections.singletonList(vector.getField()), null);
+      try (FileOutputStream fileOutputStream = new FileOutputStream(file);
+           VectorSchemaRoot root = new VectorSchemaRoot(schema, Collections.singletonList((FieldVector) vector), vector.getValueCount());
+           ArrowFileWriter writer = new ArrowFileWriter(root, new MapDictionaryProvider(), fileOutputStream.getChannel());) {
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: the second argument `provider` is usually null if there is no Dictionary being used

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java [hunk 9]
```diff
@@ -99,4 +98,24 @@ public void testReadWrite() throws IOException {
...
   }
+
+  @Test
+  public void testReadWriteMultipleBatches() throws IOException {
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
+
+    try (IntVector vector = new IntVector("foo", allocator);) {
+      Schema schema = new Schema(Collections.singletonList(vector.getField()), null);
+      try (VectorSchemaRoot root = new VectorSchemaRoot(schema, Collections.singletonList((FieldVector) vector), vector.getValueCount());
+           ArrowStreamWriter writer = new ArrowStreamWriter(root, new DictionaryProvider.MapDictionaryProvider(), Channels.newChannel(os));) {
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowStream.java]: Same thing here, `provider` can just be null and you can just put the output stream directly in the constructor like this and avoid importing Channels:
```
ArrowStreamWriter writer = new ArrowStreamWriter(root, null, out)
```

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 10]
```diff
@@ -140,22 +140,23 @@ public static int getNullCount(final ArrowBuf validityBuffer, final int valueCou
...
+    final int remainder = valueCount % 8;
 
-    for (int i = 0; i < sizeInBytes; ++i) {
-      final byte byteValue = validityBuffer.getByte(i);
-      /* Java uses two's complement binary representation, hence 11111111_b which is -1
-       * when converted to Int will have 32bits set to 1. Masking the MSB and then
-       * adding it back solves the issue.
-       */
-      count += Integer.bitCount(byteValue & 0x7F) - (byteValue >> 7);
+    for (int i = 0; i < sizeInBytes - 1; i++) {
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java]: it's minor, but to be slightly more optimized maybe avoid the `- 1` in the loop condition and do something like:
```
final int sizeInBytesMinus1 = sizeInBytes - 1;
for (int i = 0; i < sizeInBytesMinus1; ++i) {
    ...
```

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 11]
```diff
@@ -140,22 +140,23 @@ public static int getNullCount(final ArrowBuf validityBuffer, final int valueCou
...
+      byte byteValue = validityBuffer.getByte(i);
+      count += Integer.bitCount(byteValue & 0xFF);
+    }
+
+    // handling with the last byte
+    byte byteValue = validityBuffer.getByte(sizeInBytes - 1);
+    if (remainder != 0) {
+      // making the remaining bits all 1s if it is not fully filled
+      byte mask = (byte)(0xFF << remainder);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java]: nit: space after (byte)

File: java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java [hunk 12]
```diff
@@ -140,22 +140,23 @@ public static int getNullCount(final ArrowBuf validityBuffer, final int valueCou
     }
     int count = 0;
     final int sizeInBytes = getValidityBufferSize(valueCount);
+    final int remainder = valueCount % 8;
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java]: would you mind adding a comment similar to what was there before like `// If value count is not a multiple of 8, then calculate number of used bits in the last byte`

**Review by BryanCutler — COMMENTED** [review]
@bomeng I just had a few minor request if you wouldn't mind taking another look before I commit, thanks!

**Comment by bomeng** [comment]
@BryanCutler I did what you suggested in the comments, please review. Thanks.

**Comment by BryanCutler** [comment]
Thanks @bomeng , merged to master!

**Comment by BryanCutler** [comment]
@xhochy would you mind adding contributor privileges to @bomeng so I can assign the JIRA to him?  Thanks!

**Comment by xhochy** [comment]
@BryanCutler @bomeng Done. 

@bomeng you now have the necessary karma to also self-assign issues.
